core/network/fabric-interop-cc/contracts/interop/manage_assets.go

```diff
@@ -0,0 +1,304 @@
+
+	assetLockValBytes, err := ctx.GetStub().GetState(assetLockKey)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+        if assetLockValBytes != nil {
+		errorMsg := fmt.Sprintf("Asset of type %s and ID %s is already locked", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	assetLockValBytes, err = json.Marshal(assetLockVal)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Marshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+	return ctx.GetStub().PutState(assetLockKey, assetLockValBytes)
+}
+
+// UnLockAsset cc is used to record unlocking of an asset on the ledger
+func (s *SmartContract) UnLockAsset(ctx contractapi.TransactionContextInterface, assetAgreementBytes string) error {
+
+	assetAgreement := &common.AssetExchangeAgreement{}
+	err := proto.Unmarshal([]byte(assetAgreementBytes), assetAgreement)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+	//display the requested asset agreement
+	log.Info(fmt.Printf("assetExchangeAgreement: %+v\n", assetAgreement))
+
+	assetLockKey := assetAgreement.Type + ":" + assetAgreement.Id
+
+	assetLockValBytes, err := ctx.GetStub().GetState(assetLockKey)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+        if assetLockValBytes == nil {
+		errorMsg := fmt.Sprintf("No asset of type %s and ID %s is locked", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	assetLockVal := AssetLockValue{}
+	err = json.Unmarshal(assetLockValBytes, &assetLockVal)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if assetLockVal.Locker != assetAgreement.Locker || assetLockVal.Recipient != assetAgreement.Recipient {
+		errorMsg := fmt.Sprintf("Cannot unlock asset of type %s and ID %s as it is locked by %s for %s", assetAgreement.Type, assetAgreement.Id, assetLockVal.Locker, assetLockVal.Recipient)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// Check if expiry time is elapsed
+	currentTimeSecs := time.Now().UnixNano() / int64(time.Second)
+	if uint64(currentTimeSecs) < assetLockVal.ExpiryTimeSecs {
+		errorMsg := fmt.Sprintf("Cannot unlock asset of type %s and ID %s as the expiry time is not yet elapsed", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	err = ctx.GetStub().DelState(assetLockKey)
+	if err != nil {
+		errorMessage := fmt.Sprintf("Failed to delete lock for asset of type %s and ID %s: %v", assetAgreement.Type, assetAgreement.Id, err)
+		log.Error(errorMessage)
+		return errors.New(errorMessage)
+	}
+
+	return nil
+}
+
+// IsAssetLocked cc is used to query the ledger and findout if an asset is locked or not
+func (s *SmartContract) IsAssetLocked(ctx contractapi.TransactionContextInterface, assetAgreementBytes string) (bool, error) {
```

Comment:
I don't see any time checks in this function. I believe we'd determined in our design discussions that an asset's lock's status would be determined by time expiration and not solely through claims or reclaims.

I think we should add logic here whereby the function returns a false if the current time exceeds the expiry time associated with a lock. Since this is a query, we won't update or delete state, but leave it to the user to submit a Claim or an Unlock, as the case may be.